


\documentclass[a4paper]{article}
\usepackage[margin=20mm]{geometry}
\usepackage[utf8x]{inputenc}
\usepackage{xspace}
\usepackage{adjustbox}
\usepackage{tabularx}
\usepackage{longtable}

\newcommand{\GA}{an Illumina GA\xspace}
\newcommand{\nbsamples}{44\xspace}
\newcommand{\crac}{CRAC\xspace}
\newcommand{\chimct}{chimCT\xspace}
\newcommand{\rnaseq}{RNA-Seq\xspace}

\title{\textsc{Wang 2012 : Crac \& ChimCT report}}

\author{J. Audoux, A. Boureux, F. Ruffle, T. Commes and N. Philippe}
\date{\today}

\begin{document}

\maketitle
\tableofcontents

\section{Introduction}
This report describes the transcriptome analysis performed on \nbsamples RNA-seq samples 
(Illumina GA sequencing) for the identification of recurrent chimeric RNA (chRNAs) 
detected with the chimCT pipeline. ChimCT pipeline is an extension of the CRAC software 
recently published in Genome Biology by our group (Philippe et al, 2013).
Each sample is renamed using the following nomenclature:

\begin{verbatim}
AML_[normal,abnormal,unknown]_[BM,PBMC]_n[1,...,45]_[1,2].fastq 
\end{verbatim}

\begin{enumerate}
\item AML for the leukemia type
\item the karyotype (either normal,abnormal or unknown)
\item cell type (either BM or PBMC)
\item patient number from Wang et al. manuscript (2012)
\item pair: 1 = first, 2 = second
\end{enumerate}

%The sample (AML100405_2/s_3_[1,2]_sequence.txt) was removed from the analysis because the first pair file was truncated (only 511,652 reads instead of 32,673,030 reads in the second pair file).
The full list is shown on table \ref{listrenamed} (page \pageref{listrenamed}).

\section{Method}

\subsection{Step 1: filtering step and mapping reads with CRAC}
Before the mapping of RNA-Seq reads, we decided to remove reads that contain "N" to
limit noise or artefact. The table \ref{listfiltered} (page \pageref{listfiltered}) shows for each sample the number of reads 
before and after filtration. The figure \ref{fig:plotfiltered} represented the number of reads before and after
removing reads that contain “N”. For all samples, there is a low number of removed reads.
Figures \ref{fig:plotfiltered2} and \ref{fig:plotfiltered3} showed the reads distribution according to Karyotype (figure \ref{fig:plotfiltered2}) or the cell-type
(figure \ref{fig:plotfiltered3}).

CRAC was set to optimize location on 50bp paired-end reads on human genome (Philippe 
et al, 2013). For the procedure see figure \ref{fig:CRAC-chimera-workflow}, the output file of
CRAC software is a SAM file that could be used for complementary analysis. The algorithmic 
filters were indicated in the left part of the figure.

\subsection{Step 2: chimera analysis with ChimCT}

The pipeline was presented in figure \ref{fig:chimCT_workflow} and the details of the procedure are described in the
legend. In the output file, you can find the following information. 

\subsubsection{ChRNA classes}

CRAC partitions all chRNA in four classes depending on the exon organization; this partition 
resembles that depicted in [Gingeras et al, 2009]. The four classes are as follows:

\begin{enumerate}

\item The exons are located on different chromosomes

\item The exons are colinear but (likely) belong to different genes; this is checked with 
annotation.

\item The exons are on the same chromosome and same strand, but not in the order in which 
they are found on DNA, and they do not overlap each other. Of note that in class 3, there is a 
subgroup of chRNA with exons that overlap each other by at least one nucleotide.

\item The exons are on the same chromosome but on different strands.

\end{enumerate}

In \textbf{class 1}, the splicing joins pieces from distinct chromosomes, while in \textbf{classes 2 to 4} the 
exons are on the same chromosome. In summary, \textbf{class 2} is the only colinear case.
We create \textbf{class 5} to distinguish cases truely due to large scale inversions (\textbf{class 3}) from those 
likely due to local inversions or repeats inside genes. When analyzing the breast cancer 
libraries, we found many such cases.

\subsubsection{Annotation}

To investigate more closely these candidates, we annotated them to Ensembl annotations 
and could determine whether the involved "exons" are in annotated exons, introns, or 
outside genes. When there is overlapping annotation, the annotation priority is as followed: 
a higher priority to gene v/s intergenic annotation and to annotations on the same strand 
rather than to annotations on the opposite strand. The classification proceeded as follows. If 
the gene and a read are in the same orientation (sense), a read could be exonic (inxonic) or 
intronic. The same approach was used with genes on the opposite strand (if any): anti-sense 
read could thus be exonic, (inxonic), or intronic . If a read was not annotated, we assessed 
its possible intergenic localisation (NONE) (see also Philippe et al, NAR, 2013). Of note that 
for non-stranded RNA-seq experiment, the given annotation takes into account the read 
position and not the transcript strand.

\section{Results}

\subsection{Global data and recurrent chimera}

% UPDATE TABLE 1
In the present report, we focused on recurrent chRNAs identification. The identification 
process was described in the figures \ref{fig:CRAC-chimera-workflow} and \ref{fig:chimCT_workflow}. The table \ref{tab:chimCT_common-chimeras-samples-details} described the number of chRNA 
obtained for each patient and each class. As most of class 1 chRNAs are specific of a patient, 
we only reported chRNA of class 2, 3 and 4 (see method section). The chRNA number varied 
from 5 to 103 and could also depend on the reads number in the RNA-seq experiment. 
Figure \ref{fig:chimCT_class-distribution} showed the global distribution in the AML cohort. There is a higher number 
ofchRNAs in class 2 and 4 (fig \ref{fig:chimCT_class-distribution}A) and the best rank (rank =100) was observed in class 2 (fig 
\ref{fig:chimCT_class-distribution}B). This result is an agreement with our own data and the higher biological validation rate 
obtained for class2 chRNAs. The rank dispersion given in figure \ref{fig:chimCT_rank-dispersion}A and B showed a different 
between class 2 and 4 and two group of chRNA one group of rank=100 in class 2 and a group 
of rank=80 present in the two classes. Figure \ref{fig:chimCT_common-chimeras-redundancy-distribution} showed the redundancy of chRNA by giving 
the number of recurrence among the \nbsamples samples.

\subsection{Comparison with Wang et al.'s chimeras}

In the paper, a selection of 134 was selected for further biological validations. After the 
execution of CRAC and chimCT, all the corresponding reads (given in 
the supplementary data S3) were mapped and annotated on the human genome hg19. The 
134 selected chimeras are distributed as followed:

\begin{enumerate}

\item  69 of them are not considered as chimera by CRAC for the following reasons :

  \begin{enumerate}

    \item Algorithmic consideration (17 candidates), for example 13 reads correspond to 
    ambiguous candidates because of several matches on the genome

    \item Artifacts (6 candidates) as considered by Crac and ChimCT pipeline

    \item Annotation problem between the hg18 and hg19 genome version (45 candidates), 
    Among them, 5 candidates have corresponding reads that match consecutively on 
    the hg19. For 40 candidates the reads correspond to a splice junction (junction inside 
    one gene) 

    \item 1 read is a duplication of the same chimera

  \end{enumerate}

\item 65 of them are conserved (identifiées par les reads correspondants)

\end{enumerate}

Among the 134 selected chRNA, 30 of them have been validated by PCR. Among the 30 
chimeras that have been validated by Wang et al, only 11 are considered as chRNA by 
the pipeline. The main reason is that 12 of them are annotated into a splice according to 
ensembl v73. The other 7 are ruled out by CRAC because they are ambiguous or do not pass 
the stringent test.

\subsection{New results}

\subsubsection{Samples details}

Figures 8 to 10 present the result of the recurrent chRNA for each class respectively. The gene 
annotation, rank and coverage is given for each chRNA. For simplification, the ChRNA of higher rank 
superior to 85 are given. 

See figures XXX et faire un commentaire

\section{Perspectives}

**Biological validation 

We provided a file (CSV) that could be easyly open with excell

The ouput format is described as below (or on page ----), the different colomns correspond to 

15; The recurrence number is indicated

Fichier brut (tabulé) à ouvrir de voir p 10 ouput format 

15 nbre de recurrence

Email 

Qu’ils prennent connaissance des resultats dicuter 

Possiblité de validation bio ?

Ce qu’on leur conseille 

Classe 2 et Classe 4 

A différents niveaux de redondances

Plus c’est recurrent plus ? polymorphisme /artefacts tech 

Le fichier contient plsu d’infos que la figure car celle-ci tient compte du rank de 84 à 100. et 

pareil pour la couverture 

Legend des clusters sous ensemble du tableau extrait avec les critées (class, rank et support) 

**X Cliniques ? 

Frequence des chRNA 

Donner les stat de Florence pour les validations

Transcripts Network

Hypergenome Brower

Commenatire sur classe 1 et variants que l’on peut sortir avec les spanning reads

%\crac was set to optimize location on 50bp paired-end reads on human
%genome. Plus, two options have been used to increase the precision in
%the chimera detection: ``--no-ambiguity'' to discard ambiguous
%chimeras and ``--stringent-chimera'' to add some criteria in
%algorithmic verification.
%
%\section{Step 3: chimera analysis with ChimCT}
%
%\subsection{Comparison with Wang \textit{et al.}'s chimeras} 
%In the paper, a selection of 134 were tested and 30 of them have been
%validated. After the execution of \crac et \chimct on the \nbsamples
%patients, the 134 chimeras are distributed as follow:
%\begin{enumerate}
%\item 28 of them are not considered as chimera by \crac:
%  \begin{itemize}
%  \item 5 match consecutively on the GRCh37 genome (hg19)
%  \item 4 do not pass the stringent test (--stringent
%    argument)
%  \item 13 are ambiguous because of several matches on the genome
%  \item 6 are considered artifacts
%  \end{itemize}
%\item 40 of them are not considered as chimera by \chimct because they
%  correspond to a splice junction (junction inside once gene) 
%\item 1 is a duplication of the same chimera 
%\item {\bf 65 of them are conserved}
%\end{enumerate}
%
%Among the 30 chimeras that have been validated by Wang \textit{et
%al.}. , only 11 are considered in our pipeline. The main reason is
%that 12 of them are annotated into a splice according to ensembl
%v73. The other 7 are ruled out by \crac because they are ambiguous or
%do not pass the stringent test.
%
%Ultimately, our pipeline found 65 different chimeras and 66 chimeras
%among the 134 but 45 of them no longer represent chimeras because of
%an update of the annotations on human genome hg19. As for the last 23,
%they are considered ambiguous or artefactual in our pipeline. Note
%that the use of short reads makes the rigor of \crac in order to keep
%high accuracy, while using longer reads would also increase the
%sensitivity.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%% FILTERING FIGURES
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% latex table generated in R 3.0.0 by xtable 1.7-1 package
% Tue Dec  3 16:32:42 2013
\begin{table}[ht]
\centering
\begin{tabular}{rlll}
  \hline
 & old\_filename & new\_filenames & nb\_reads \\ 
  \hline
1 & AML100405\_2/s\_1\_[1,2]\_sequence.txt & AML\_normal\_BM\_n1\_[1,2].fastq & 2X24614703 \\ 
  2 & AML100405\_2/s\_2\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n7\_[1,2].fastq & 2X30770492 \\ 
  3 & AML100405\_2/s\_4\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n14\_[1,2].fastq & 2X32629136 \\ 
  4 & AML100405\_2/s\_5\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n15\_[1,2].fastq & 2X27996366 \\ 
  5 & AML100405\_2/s\_6\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n16\_[1,2].fastq & 2X30014538 \\ 
  6 & AML100405\_2/s\_7\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n17\_[1,2].fastq & 2X31305698 \\ 
  7 & AML100405\_2/s\_8\_[1,2]\_sequence.txt & AML\_normal\_BM\_n5\_[1,2].fastq & 2X27914139 \\ 
  8 & AML100413\_3/s\_1\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n18\_[1,2].fastq & 2X27985862 \\ 
  9 & AML100413\_3/s\_2\_[1,2]\_sequence.txt & AML\_normal\_BM\_n6\_[1,2].fastq & 2X15830586 \\ 
  10 & AML100413\_3/s\_3\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n19\_[1,2].fastq & 2X33253655 \\ 
  11 & AML100413\_3/s\_4\_[1,2]\_sequence.txt & AML\_unknown\_BM\_n38\_[1,2].fastq & 2X32368840 \\ 
  12 & AML100413\_3/s\_5\_[1,2]\_sequence.txt & AML\_unknown\_BM\_n39\_[1,2].fastq & 2X22547358 \\ 
  13 & AML100413\_3/s\_6\_[1,2]\_sequence.txt & AML\_unknown\_PBMC\_n43\_[1,2].fastq & 2X8036353 \\ 
  14 & AML100413\_3/s\_7\_[1,2]\_sequence.txt & AML\_unknown\_PBMC\_n44\_[1,2].fastq & 2X20728171 \\ 
  15 & AML100413\_3/s\_8\_[1,2]\_sequence.txt & AML\_unknown\_BM\_n40\_[1,2].fastq & 2X27173669 \\ 
  16 & AML100713\_5/s\_1\_[1,2]\_sequence.txt & AML\_normal\_BM\_n2\_[1,2].fastq & 2X28843603 \\ 
  17 & AML100713\_5/s\_2\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n8\_[1,2].fastq & 2X28597393 \\ 
  18 & AML100713\_5/s\_3\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n9\_[1,2].fastq & 2X29105962 \\ 
  19 & AML100713\_5/s\_4\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n10\_[1,2].fastq & 2X32766978 \\ 
  20 & AML100713\_5/s\_5\_[1,2]\_sequence.txt & AML\_normal\_BM\_n3\_[1,2].fastq & 2X31267542 \\ 
  21 & AML100713\_5/s\_6\_[1,2]\_sequence.txt & AML\_normal\_BM\_n4\_[1,2].fastq & 2X32518678 \\ 
  22 & AML100713\_5/s\_7\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n11\_[1,2].fastq & 2X28074606 \\ 
  23 & AML100713\_5/s\_8\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n12\_[1,2].fastq & 2X27129917 \\ 
  24 & AML101130/s\_1\_[1,2]\_sequence.txt & AML\_normal\_BM\_n21\_[1,2].fastq & 2X34693529 \\ 
  25 & AML101130/s\_2\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n27\_[1,2].fastq & 2X35331090 \\ 
  26 & AML101130/s\_3\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n34\_[1,2].fastq & 2X35780585 \\ 
  27 & AML101130/s\_5\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n26\_[1,2].fastq & 2X37941973 \\ 
  28 & AML101130/s\_6\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n30\_[1,2].fastq & 2X34785818 \\ 
  29 & AML101130/s\_7\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n31\_[1,2].fastq & 2X35756006 \\ 
  30 & AML101130/s\_8\_[1,2]\_sequence.txt & AML\_unknown\_PBMC\_n45\_[1,2].fastq & 2X16064143 \\ 
  31 & AML101209/s\_1\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n28\_[1,2].fastq & 2X24535864 \\ 
  32 & AML101209/s\_2\_[1,2]\_sequence.txt & AML\_normal\_BM\_n20\_[1,2].fastq & 2X21676424 \\ 
  33 & AML101209/s\_3\_[1,2]\_sequence.txt & AML\_normal\_PBMC\_n29\_[1,2].fastq & 2X22140260 \\ 
  34 & AML101209/s\_5\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n32\_[1,2].fastq & 2X16978603 \\ 
  35 & AML101209/s\_6\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n36\_[1,2].fastq & 2X29058205 \\ 
  36 & AML101209/s\_7\_[1,2]\_sequence.txt & AML\_normal\_BM\_n22\_[1,2].fastq & 2X20786654 \\ 
  37 & AML101209/s\_8\_[1,2]\_sequence.txt & AML\_normal\_BM\_n23\_[1,2].fastq & 2X26504532 \\ 
  38 & AML110509/s\_1\_[1,2]\_sequence.txt & AML\_unknown\_BM\_n41\_[1,2].fastq & 2X44459470 \\ 
  39 & AML110509/s\_2\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n33\_[1,2].fastq & 2X46317335 \\ 
  40 & AML110509/s\_3\_[1,2]\_sequence.txt & AML\_normal\_BM\_n24\_[1,2].fastq & 2X44998831 \\ 
  41 & AML110509/s\_5\_[1,2]\_sequence.txt & AML\_normal\_BM\_n25\_[1,2].fastq & 2X44645304 \\ 
  42 & AML110509/s\_6\_[1,2]\_sequence.txt & AML\_abnormal\_BM\_n35\_[1,2].fastq & 2X46815659 \\ 
  43 & AML110509/s\_7\_[1,2]\_sequence.txt & AML\_unknown\_BM\_n42\_[1,2].fastq & 2X46567887 \\ 
  44 & AML110509/s\_8\_[1,2]\_sequence.txt & AML\_abnormal\_PBMC\_n37\_[1,2].fastq & 2X44234187 \\ 
   \hline
\end{tabular}
\caption{Old \& New filename for the \nbsamples AML samples.} 
\label{listrenamed}
\end{table}



% latex table generated in R 3.0.0 by xtable 1.7-1 package
% Tue Dec  3 16:32:42 2013
\begin{table}[ht]
\centering
\begin{tabular}{rlllll}
  \hline
 & Patients & nb\_reads & reads\_length & nb\_filtered & nb\_remained \\ 
  \hline
1 & AML normal BM n1 & 24614703 & 49 & 558197 & 24056506 \\ 
  2 & AML normal PBMC n7 & 30770492 & 49 & 403037 & 30367455 \\ 
  3 & AML normal PBMC n14 & 32629136 & 49 & 196545 & 32432591 \\ 
  4 & AML normal PBMC n15 & 27996366 & 49 & 173027 & 27823339 \\ 
  5 & AML normal PBMC n16 & 30014538 & 49 & 180665 & 29833873 \\ 
  6 & AML normal PBMC n17 & 31305698 & 49 & 192793 & 31112905 \\ 
  7 & AML normal BM n5 & 27914139 & 49 & 211278 & 27702924 \\ 
  8 & AML normal PBMC n18 & 27985862 & 49 & 455044 & 27530818 \\ 
  9 & AML normal BM n6 & 15830586 & 49 & 23602 & 15806984 \\ 
  10 & AML normal PBMC n19 & 33253655 & 49 & 114989 & 33138666 \\ 
  11 & AML unknown BM n38 & 32368840 & 49 & 54853 & 32313987 \\ 
  12 & AML unknown BM n39 & 22547358 & 49 & 539706 & 22007652 \\ 
  13 & AML unknown PBMC n43 & 8036353 & 49 & 1712570 & 6323783 \\ 
  14 & AML unknown PBMC n44 & 20728171 & 49 & 280434 & 20447737 \\ 
  15 & AML unknown BM n40 & 27173669 & 49 & 67869 & 27105800 \\ 
  16 & AML normal BM n2 & 28843603 & 50 & 921900 & 27921703 \\ 
  17 & AML normal PBMC n8 & 28597393 & 50 & 192433 & 28404960 \\ 
  18 & AML normal PBMC n9 & 29105962 & 50 & 195417 & 28910545 \\ 
  19 & AML normal PBMC n10 & 32766978 & 50 & 210134 & 32556844 \\ 
  20 & AML normal BM n3 & 31267542 & 50 & 202139 & 31065403 \\ 
  21 & AML normal BM n4 & 32518678 & 50 & 210417 & 32308261 \\ 
  22 & AML normal PBMC n11 & 28074606 & 50 & 184095 & 27890511 \\ 
  23 & AML normal PBMC n12 & 27129917 & 50 & 177553 & 26952364 \\ 
  24 & AML normal BM n21 & 34693529 & 50 & 5597158 & 29096371 \\ 
  25 & AML normal PBMC n27 & 35331090 & 50 & 1459837 & 33871253 \\ 
  26 & AML abnormal BM n34 & 35780585 & 50 & 1122974 & 34657611 \\ 
  27 & AML normal PBMC n26 & 37941973 & 50 & 199463 & 37742510 \\ 
  28 & AML abnormal BM n30 & 34785818 & 50 & 456952 & 34328866 \\ 
  29 & AML abnormal BM n31 & 35756006 & 50 & 189027 & 35566979 \\ 
  30 & AML unknown PBMC n45 & 16064143 & 50 & 86696 & 15977447 \\ 
  31 & AML normal PBMC n28 & 24535864 & 50 & 1698402 & 22837462 \\ 
  32 & AML normal BM n20 & 21676424 & 50 & 85157 & 21591267 \\ 
  33 & AML normal PBMC n29 & 22140260 & 50 & 551348 & 21588912 \\ 
  34 & AML abnormal BM n32 & 16978603 & 50 & 68933 & 16909670 \\ 
  35 & AML abnormal BM n36 & 29058205 & 50 & 106481 & 28951724 \\ 
  36 & AML normal BM n22 & 20786654 & 50 & 77044 & 20709610 \\ 
  37 & AML normal BM n23 & 26504532 & 50 & 118625 & 26385907 \\ 
  38 & AML unknown BM n41 & 44459470 & 50 & 1496273 & 42963197 \\ 
  39 & AML abnormal BM n33 & 46317335 & 50 & 2683806 & 43633529 \\ 
  40 & AML normal BM n24 & 44998831 & 50 & 1143562 & 43855269 \\ 
  41 & AML normal BM n25 & 44645304 & 50 & 52577 & 44592727 \\ 
  42 & AML abnormal BM n35 & 46815659 & 50 & 54280 & 46761379 \\ 
  43 & AML unknown BM n42 & 46567887 & 50 & 436383 & 46131504 \\ 
  44 & AML abnormal PBMC n37 & 44234187 & 50 & 413760 & 43820427 \\ 
   \hline
\end{tabular}
\caption{Number or reads filtered for each pair of files} 
\label{listfiltered}
\end{table}



\begin{figure}[]

\includegraphics[width=\maxwidth]{figure/plotfiltered} \caption[Number of reads before and after filtration]{Number of reads before and after filtration\label{fig:plotfiltered}}
\end{figure}




\begin{figure}[]


{\centering \includegraphics[width=\maxwidth]{figure/plotfiltered2} 

}

\caption[Boxplots of reads filtration by karyotype]{Boxplots of reads filtration by karyotype\label{fig:plotfiltered2}}
\end{figure}




\begin{figure}[]


{\centering \includegraphics[width=\maxwidth]{figure/plotfiltered3} 

}

\caption[Boxplots of reads filtration by cell type]{Boxplots of reads filtration by cell type\label{fig:plotfiltered3}}
\end{figure}







\subsubsection{Samples details}

The table below presents the samples details that have be used to perform the search of redundant chimeras.

% latex table generated in R 3.0.0 by xtable 1.7-1 package
% Tue Dec  3 16:32:44 2013
\begin{longtable}{rlrrrrr}
  \hline
 & Sample & 2 & 3 & 4 & nb\_redundant\_chimeras & cover \\ 
  \hline
1 & AML\_abnormal\_BM\_n30 &  38 &   2 &  36 &  76 & 203366 \\ 
  2 & AML\_abnormal\_BM\_n31 &  43 &   4 &  38 &  85 & 402595 \\ 
  3 & AML\_abnormal\_BM\_n32 &  22 &   2 &  48 &  72 & 955320 \\ 
  4 & AML\_abnormal\_BM\_n33 &  45 &   2 &   1 &  48 & 101308 \\ 
  5 & AML\_abnormal\_BM\_n34 &  32 &   4 &  54 &  90 & 431065 \\ 
  6 & AML\_abnormal\_BM\_n35 &  47 &   4 &   4 &  55 & 309228 \\ 
  7 & AML\_abnormal\_BM\_n36 &  29 &   5 &  55 &  89 & 2803483 \\ 
  8 & AML\_abnormal\_PBMC\_n37 &  19 &   2 &   5 &  26 & 282514 \\ 
  9 & AML\_normal\_BM\_n1 &  10 &   0 &  23 &  33 & 54983 \\ 
  10 & AML\_normal\_BM\_n2 &  20 &   4 &  16 &  40 & 727486 \\ 
  11 & AML\_normal\_BM\_n20 &  18 &   2 &  31 &  51 & 655818 \\ 
  12 & AML\_normal\_BM\_n21 &  29 &   1 &  73 & 103 & 288123 \\ 
  13 & AML\_normal\_BM\_n22 &  14 &   2 &  61 &  77 & 2444750 \\ 
  14 & AML\_normal\_BM\_n23 &  19 &   1 &  48 &  68 & 803615 \\ 
  15 & AML\_normal\_BM\_n24 &  44 &   5 &   8 &  57 & 202063 \\ 
  16 & AML\_normal\_BM\_n25 &  34 &   4 &   2 &  40 & 961517 \\ 
  17 & AML\_normal\_BM\_n3 &  24 &   3 &   7 &  34 & 298092 \\ 
  18 & AML\_normal\_BM\_n4 &  26 &   1 &   8 &  35 & 151933 \\ 
  19 & AML\_normal\_BM\_n5 &   9 &   2 &   1 &  12 & 14742 \\ 
  20 & AML\_normal\_BM\_n6 &   9 &   1 &   3 &  13 & 58138 \\ 
  21 & AML\_normal\_PBMC\_n10 &  29 &   3 &   6 &  38 & 413105 \\ 
  22 & AML\_normal\_PBMC\_n11 &  34 &   1 &   6 &  41 & 31056 \\ 
  23 & AML\_normal\_PBMC\_n12 &   7 &   1 &   8 &  16 & 79236 \\ 
  24 & AML\_normal\_PBMC\_n13 &   2 &   2 &   1 &   5 & 60434 \\ 
  25 & AML\_normal\_PBMC\_n14 &  21 &   3 &   2 &  26 & 52689 \\ 
  26 & AML\_normal\_PBMC\_n15 &  24 &   3 &   1 &  28 & 193728 \\ 
  27 & AML\_normal\_PBMC\_n16 &  21 &   1 &   2 &  24 & 15191 \\ 
  28 & AML\_normal\_PBMC\_n17 &  19 &   2 &   3 &  24 & 61802 \\ 
  29 & AML\_normal\_PBMC\_n18 &  21 &   2 &   1 &  24 & 67701 \\ 
  30 & AML\_normal\_PBMC\_n19 &  24 &   1 &  13 &  38 & 144102 \\ 
  31 & AML\_normal\_PBMC\_n26 &  37 &   2 &  45 &  84 & 234808 \\ 
  32 & AML\_normal\_PBMC\_n27 &  36 &   2 &  56 &  94 & 594114 \\ 
  33 & AML\_normal\_PBMC\_n28 &  33 &   2 &  42 &  77 & 412006 \\ 
  34 & AML\_normal\_PBMC\_n29 &  21 &   2 &  41 &  64 & 640571 \\ 
  35 & AML\_normal\_PBMC\_n7 &  21 &   1 &   6 &  28 & 17510 \\ 
  36 & AML\_normal\_PBMC\_n8 &  15 &   5 &   4 &  24 & 354710 \\ 
  37 & AML\_normal\_PBMC\_n9 &  15 &   2 &   7 &  24 & 208074 \\ 
  38 & AML\_unknown\_BM\_n38 &  15 &   4 &  10 &  29 & 423110 \\ 
  39 & AML\_unknown\_BM\_n39 &   6 &   3 &  11 &  20 & 129320 \\ 
  40 & AML\_unknown\_BM\_n40 &  16 &   1 &  15 &  32 & 109565 \\ 
  41 & AML\_unknown\_BM\_n41 &  24 &   3 &   2 &  29 & 364344 \\ 
  42 & AML\_unknown\_BM\_n42 &  53 &   5 &   3 &  61 & 122021 \\ 
  43 & AML\_unknown\_PBMC\_n43 &   3 &   1 &   2 &   6 & 42063 \\ 
  44 & AML\_unknown\_PBMC\_n44 &  12 &   0 &   5 &  17 & 79273 \\ 
  45 & AML\_unknown\_PBMC\_n45 &  18 &   2 &  20 &  40 & 456081 \\ 
  46 & wang2012-chimeras-stringent-chimera &  30 &   0 &   0 &  30 & 6000 \\ 
   \hline
\hline
\caption{Summary of chimera analysis for each sample.} 
\label{tab:chimCT_common-chimeras-samples-details}
\end{longtable}




\subsubsection{General statistics}

\large{Class distribution}\\

\begin{figure}
\begin{minipage}[b]{0.45\linewidth}
\centering

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}

{\centering \includegraphics[width=\maxwidth]{figure/allChimerasClassDistribution} 

}



\end{knitrout}


\end{minipage}
\hspace{0.5cm}
\begin{minipage}[b]{0.45\linewidth}
\centering

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}

{\centering \includegraphics[width=\maxwidth]{figure/bestChimerasClassDistribution} 

}



\end{knitrout}

\end{minipage}
\caption{Chimeras distribution by class}
\label{fig:chimCT_class-distribution}
\end{figure}


\large{\bf Ranks dispersion}\\

\begin{figure}
\begin{minipage}[b]{0.45\linewidth}
\centering
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}
\includegraphics[width=\maxwidth]{figure/boxplotRankDispersion} 

\end{knitrout}

\end{minipage}
\hspace{0.5cm}
\begin{minipage}[b]{0.45\linewidth}
\centering
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}
\includegraphics[width=\maxwidth]{figure/violinRankDispersion} 

\end{knitrout}

\end{minipage}
\caption{Rank dispersion by class}
\label{fig:chimCT_rank-dispersion}
\end{figure}


\large{\bf Redundacy distribution}\\

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{figure}[]


{\centering \includegraphics[width=\maxwidth]{figure/chimCT_common-chimeras-redundancy-distribution} 

}

\caption[Redundancy distribution for each class]{Redundancy distribution for each class\label{fig:chimCT_common-chimeras-redundancy-distribution}}
\end{figure}


\end{knitrout}



\subsubsection{Redundancy heatmap}








